src/gribtonc/gribdump.1
'\"
'\" $Id: gribdump.1,v 1.3 2003/02/21 21:37:46 rkambic Exp $
.TH GRIBDUMP 1 "8 June 1995"
.SH NAME
gribdump - Decode GRIB products and display as text
.SH SYNOPSIS
.HP
.ft B
gribdump
.nh
\%[-h]
\%[-b]
\%[-p\ \fIprecision\fP]
\%[-e\ \fIerrpath\fP]
\%[-g\ \fItabfile\fP]
\%[-q\ \fIqspecs\fP]
\%[\fIfiles ...\fP]
.hy
.ft
.SH DESCRIPTION
.LP
This program decodes GRIB products (embedded in WMO product envelopes) on
the NWS High-Resolution Data Service (HRS) feed, writing textual
representations of the products or summaries of the products to standard
output.
.SH OPTIONS 
.TP
.B \-h
Headers only.  Information from the Product Description Section and Grid
Description Section is output, but no data values are output.  The default
is to output both headers and data values.
.TP
.B \-b
Brief output.  Only one line is output for each product, providing summary
information of the parameter, grid, model, levels, time ranges, and other
information.
.TP
.BI \-p " precision"
Precision used for printing floating-point values, if other than the
7-significant figure default is desired.
.TP
.BI \-e " errpath"
If a GRIB product cannot be decoded, perhaps because it is truncated or
otherwise corrupted, it is appended to this file.

.TP
.BI \-g " tabfile"
Use GRIB-netcdf translation defined in this file. The format of this file
is:
.LP
.LP
.nf
Grib;netCDF;long name               ;units;comments
 31 ;CI    ;Sea-ice cover           ;none ;(0-1)
 34 ;SSTK  ;Sea surface temperature ;K    ;
151 ;MSL   ;Mean sea-level pressure ;Pa   ;
167 ;T2M   ;2 metre temperature     ;K    ;
.fi
.RE
.TP
.BI \-q " qspecs"
Specification for how to expand so-called quasi-regular or thinned grids
into rectangular grids.  This argument is a
comma-delimited text string that specifies the method of interpolation used
(currently \fBlin\fP for linear or \fBcub\fP for cubic), and the latitude
and longitude
increments in degrees for the resulting grid (e.g. \fBdlat=2.5,dlon=5.0\fP).
.TP
.I files
The path name of one or more files of GRIB products used as input.
If no filenames are provided, GRIB products are read from standard input.
.SH EXAMPLE
The command
.RS +4
.nf
gribdump -q "lin,dlat=2.5,dlon=5.0" avn-x.wmo
.fi
.RE
will decode and display values for the products in avn-x.wmo If the data
contain quasi-regular grids, they will be expanded to full regular grids on
a 2.5 by 5.0 degree grid before being written into the output file.
.SH "SEE ALSO"
gribtonc(1), netcdf(3), pqact(1), udunits(3), ulog(3)
.SH DIAGNOSTICS
Error messages and log messages are written to the log file.



src/RELEASE-NOTES
$Id: RELEASE-NOTES,v 1.37 2004/12/21 20:55:26 rkambic Exp $

Release	Date		Notes
_________________	______________________________________________________
5.0.2 2013-05-07	Corrected gfs_37-44.cdl (grids are 1.25x1.25 degree);
                        added ukmet_21-24.cdl (like GFS AVN, these are 1.25x1.25 degree
                        grids); added '-n' flag to gribtonc (says to not expand
                        the record dimension past what is defined in the CDL/existing
                        netCDF file); updated 'port' subdirectory with latest code
                        (so that the package will build on CentOS 6.4 x86_64)
5.0.1 2012-06-13	Remade ruc2_236.cdl because of transfer of the RUC2 production
			from FSL to RAP 
5.0.0 2007-10-10	The perl Decoders have been moved to their own package
			so the process of deprecating this package can take place.
			This package is being deprecated because of it's 
			limitation of not being able to decode Grib2, compile
			problems, and static parameter table.
4.1.7 2007-09-27	metar2nc decoders removes reports with illegal cntl chars
			changed dist from Z to gz format
4.1.6 2007-08-20	fixed typo in eta_211.cdl
4.1.5 2007-02-09	added a -c option to metar2nc* so it can process archive
                        data from the command line.
4.1.4 2006-10-04	fixed bug in ua2nc, putting 23hr data into wrong file.
4.1.3 2006-09-14	fixed bug in ua2nc, putting 23hr data into wrong file.
4.1.2 2006-09-06	added more levels
4.1.1 2006-08-23	minor bug fixes, recs.c added import, metar2nc.new1
			SPD and GUST are now floats when converted from knots
			to m/s. etc
4.0.0 2005-12-19	update the METAR table sfmetar_sa.tbl, created new
			metar.cdl.new1, new table lsfstns.tbl for synoptic 
			decoder, and new table for upperair decoder snstns.tbl
			updated other cdls. the new table names cause mods to
			syn2nc and ua2nc.
4.0.0 2005-10-05	changed perl from /usr/local/bin/perl to /usr/bin/perl
			pqact entries don't need yyyymm argument anymore.
			new naming convention, ie surface_METAR_yyyymmdd_hhmm.nc
			added -h option to write hourly files for the all perl
			decoders, default is daily file. updated sfmetar.tbl, 
			snworld.tbl and systns_upc.tbl. fixed bug in syn2nc 
			that returned wind DIR greater than 900, removed java 
			Grib decoder jars, it's has it own release package now
			called grib-decoder-java_2.2.10.tar.gz. gfs_211.cdl,
			eta_211.cdl, and ruc2_236.cdl were updated to add
			new parmeters. managed log files to only keeps 3 logs
			added global attributes to metar.cdl.new1
			changed testdata.wmo to testdata.wmo.bak, etc

3.1.1 2005-05-24	changed units of Dx and Dy to be meters instead of km 
			in all cdls. changed cdl version to be 1.0
3.1.0 2005-05-24	fixed the perl decoders so unicode characters don't put
			decoder into infinite loop. This fix is needed for 
			perl version 5.8.0 and higher. fixed buoy2nc and syn2nc
			where min was set to 60 and timegm() was called. 
			Also added the java grib.jar to the release.

3.0.9 2005-03-01	fixed the perl decoder cdls to be udunits compliant.  
3.0.8 2005-02-04	fixed metar2nc and metar2nc.new1 dew point temperature
			incase the coding changes from /M00 to /M0
3.0.7 2004-12-21	added level 245, changed files: levels.h gribtocdl.c 
			levels.c nc.c
3.0.6 2004-10-21	some parameters started with a number, thats illegal
			in netcdf. also correct units for oztot (ozone total)
3.0.5 2004-10-18	update cdls and new version of metar.new1, metar.cdl.new

3.0.4 2004-07-22	rewrote perl decoders to be more robust especially
			when checking the report time for accuracy. Changed
			test to read in more stations from the sfmetar_sa.tbl
			station file. Added metar2nc.new1 and metar.cdl.new1 as
			a new perl decoder, new netCDF format. New script
			makeCDL1.pl makes a cdl file metar.cdl.new1. In recs.c
			upped RECS_INIT_SIZE to 64. Updated cdls
			In nc.c added stronger test to parmname, levels
			mods to test.pl for real time data, mods to
			src/*/testdata.wmo for real time testing. Fixed 
			return value for time range indicator in gribtocdl

3.0.3 2004-03-12	added Writer scripts into all the perl decoders. These
			scripts bust-up bulletins into individual reports
			added missing level info to gribtocdl.c program
			added Makefile.in to zone MANIFEST
			Fixed century bug, files recs.c units.c gribdump

3.0.2 2004-02-01	change #define MAX_GRIB_SIZE 200000
			added forecasttime variable to gribtocdl
			rename all cdls to reflect new model names
			updated INSTALL and CUSTOMIZE files
			in cdl dir, updated Makefile.in and README
			mods to metar2nc, made matches more robust

3.0.1 2003-04-25	added CF standard_name, forecasttime, and enhanced
			Conventions attribute to acknowledge CF conventions
			to all model cdls. Modified recs.c to write out
			forecasttime.

3.0.0 2003-03-25	added dimension valtime_offset for the valtime_offset
			variable. Enhanced cdl to have the data for the 
			valtime_offset var.  This permits the data to be stored
			in monotonic increasing order.  Added parameter_number
			and level_number attributes to variables.

2.4.4 2003-02-25        added vars gust_sfc to ruc2.cdl and vars Tmin_fhg and
			Tmax_fhg to avn-x.cdl
			mods to gdes.c gribtocdl.[1c] inetutil.c ulog.h so 
			code compiles on AIX boxes.

2.4.3 2003-02-21        modifications so user parms functionality works
			correctly in the code, new functionality to 
			gribdump, also updated the man pages. files:
			gribtocdl.c, params.c, gribdump.c, gribtonc.1 and
			gribdump.1
			
2.4.2 2003-02-19        Added the PNO indicator for Metar decoder
			In gdes.c, changed i_dim from lat to lon
			fixed avn-1.0x1.0.cdl also

2.4.1 2003-01-31        Fixed some error for the avn-1.0x1.0 Grid.  Also fixed
			errors for the Gausian lat/lon grid, plus other fixes
			mods to gribtocld.c gdes.c nc.c 

2.4.0 2002-11-04        Merge in code made by Arild Burud and Terje Reite for

src/gribtonc/gribdump.c
/*
 *	Copyright 1993, University Corporation for Atmospheric Research.
 */
/* $Id: gribdump.c,v 1.24 2004/05/26 20:57:17 rkambic Exp $ */

/* 
 * decodes GRIB products into ASCII
 */

#include <stdio.h>
#include <signal.h>
#include <errno.h>
#include <unistd.h>
#include <limits.h>		/* _POSIX_PATH_MAX */
#include <string.h>
#include <stdlib.h>
#include <math.h>
#include "ulog.h"
#include "mkdirs_open.h"

#include "centers.h"
#include "models.h"
#include "params.h"
#include "levels.h"
#include "timeunits.h"
#include "grib1.h"
#include "product_data.h"
#include "quasi.h"
#include "user_param.h"

#ifdef NO_ATEXIT
#include <atexit.h>
#endif

#ifdef __STDC__
static void cleanup(void);
static void signal_handler(int sig);
static void set_sigactions(void);
static void usage(char* av0);
static void print_floats(float* ff, int cols, int rows, int pp);
static int is_ixg(int id);
static void print_grib(struct product_data* gp, int prec);
static void print_grib_line(struct product_data* gp);
static int do_dump(FILE * fp, FILE * ep, int timeout, int prec, quas*);
static product_data* grib_decode(prod*, quas*);
int main(int ac, char** av);
#endif

unsigned long num_wmo_messages; /* for statistics on exit */
unsigned long num_gribs_decoded;
unsigned long num_gribs_written;
unsigned long num_degribbed;

int testNoserc;

/*
 * Timeout in seconds.  If no input is received for this interval, process
 * closes output file and exits.
 */
#define DEFAULT_TIMEOUT  600

/*
 * Default precision used for printing floats.
 */
#define DEFAULT_PRECISION	7

/*
 * Called at exit.
 * This callback routine registered by atexit().
 */
static void
cleanup()
{
    uinfo("Exiting") ;
    uinfo("%lu WMO msgs, %lu GRIBs decoded, %lu written ",
	  num_wmo_messages, num_gribs_decoded, num_gribs_written);
    (void) closeulog();
}


/*
 * Called upon receipt of signals.
 * This callback routine registered in set_sigactions() .
 */
static void
signal_handler(sig)
     int sig ;
{
#ifdef SVR3SIGNALS
    /* 
     * Some systems reset handler to SIG_DFL upon entry to handler.
     * In that case, we reregister our handler.
     */
    (void) signal(sig, signal_handler) ;
#endif
    switch(sig) {
      case SIGHUP :
	udebug("SIGHUP") ;
	return ;
      case SIGINT :
	unotice("Interrupt") ;
	exit(0) ;
      case SIGTERM :
	udebug("SIGTERM") ;
	exit(0) ;
      case SIGUSR1 :
	udebug("SIGUSR1") ;
	return ;
      case SIGUSR2 :
	if (toggleulogpri(LOG_INFO))
	  unotice("Going verbose") ;
	else
	  unotice("Going silent") ;
	return ;
      case SIGPIPE :
	unotice("SIGPIPE") ;
	exit(0) ;
    }
    udebug("signal_handler: unhandled signal: %d", sig) ;
}


static void
set_sigactions()
{
#ifndef NO_POSIXSIGNALS
    struct sigaction sigact ;
    
    sigact.sa_handler = signal_handler;
    sigemptyset(&sigact.sa_mask) ;
    sigact.sa_flags = 0 ;
    
    (void) sigaction(SIGHUP, &sigact, NULL) ;
    (void) sigaction(SIGINT, &sigact, NULL) ;
    (void) sigaction(SIGTERM, &sigact, NULL) ;
    (void) sigaction(SIGUSR1, &sigact, NULL) ;
    (void) sigaction(SIGUSR2, &sigact, NULL) ;
    (void) sigaction(SIGPIPE, &sigact, NULL) ;
#else
    (void) signal(SIGHUP, signal_handler) ;
    (void) signal(SIGINT, signal_handler) ;
    (void) signal(SIGTERM, signal_handler) ;
    (void) signal(SIGUSR1, signal_handler) ;
    (void) signal(SIGUSR2, signal_handler) ;
    (void) signal(SIGPIPE, signal_handler) ;
#endif
}


static void
usage(av0)
     char *av0 ; /*  id string */
{
    (void)fprintf(stderr,
		  "Usage: %s [options] [files ...]\t\nOptions:\n", av0);
    (void)fprintf(stderr,
		  "\t-v           Verbose, report decoding steps\n") ;
    (void)fprintf(stderr,
		  "\t-l logfile   Send log info to file (default uses stdout)\n") ;
    (void)fprintf(stderr,
		  "\t-h           Output headers only, no data values\n") ;
    (void)fprintf(stderr,
		  "\t-b           Brief output of headers data\n") ;
    (void)fprintf(stderr,
		  "\t-t timeout   If no input, exit after \"timeout\" seconds (default %d)\n",
		  DEFAULT_TIMEOUT) ;
    (void)fprintf(stderr,
		  "\t-e errfile   Append bad GRIB products to this file\n") ;
    (void)fprintf(stderr,
		  "\t-p prec      Precision used to print floats (default %d)\n",
		  DEFAULT_PRECISION) ;
    (void)fprintf(stderr,
		  "\t-q meth      Specs for expanding quasi-regular grids,\n\t\t\te.g. -q \"lin,dlat=2.5,dlon=5.0\"\n") ;
    exit(1);
}

int headers_only;
int brief_output;

static void
print_floats(ff, cols, rows, pp)
     float *ff;			/* array of floats to print */
     int cols; 			/* number of columns of data */
     int rows; 			/* number of rows of data */
     int pp;			/* precision to use in printing */
{
    int ii;
    int jj;
    int per_line = 8;
    static char format[] = "%.10g%s";
    
    sprintf(format, "%%.%dg%%s", pp);

    for(ii = 0; ii < rows; ii++) {
	printf("Row %d:\n\t",ii);

	for(jj = 0; jj < cols;jj++)
	    printf(format, ff[ii*cols+jj],
		   jj%per_line+1 == per_line ? "\n\t" : " ");

	if ((jj-1)%per_line+1 != per_line)
	    printf("\n");
    }
}


static int
is_ixg(id)			/* is an international exchange grid? */
     int id;
{
    switch (id) {
      case 21:
      case 22:
      case 23:
      case 24:
      case 25:
      case 26:
      case 37:
      case 38:
      case 39:
      case 40:
      case 41:
      case 42:
      case 43:
      case 44:
      case 50:
      case 61:
      case 62:
      case 63:
      case 64:
	return 1;
    }
    /* default: */
    return 0;
}

static void
print_grib(gp, prec)
     struct product_data *gp;
     int prec;			/* precision to use for printing float vals */
{

    char format[80];

    printf("-----------------------------------------------------\n");

    printf("   %24s : %s\n","    Header", gp->header);
    printf("   %24s : %d (%s)\n","Originating Center", gp->center,
	   centername(gp->center));
    if(gp->subcenter != 0)
    printf("   %33s : %d (%s)\n","Subcenter", gp->subcenter,
	   subcentername(gp->center, gp->subcenter));
    printf("   %24s : %d (%s)\n","Process", gp->model,
	   modelname(gp->center,gp->model));
    printf("   %24s : %d\n","Grid", gp->grid);
    printf("             points in grid : %d\n", gp->npts);
    printf("   %24s : %d (%s)\n","Parameter", gp->param,
	   grib_pname(gp->param));
    printf("   %24s : %s\n", "Units", grib_units(gp->param));
    printf("   %24s : %s\n", "Level Type", levelname(gp->level_flg));
    switch (gp->level_flg) {
      case LEVEL_SURFACE:
      case LEVEL_CLOUD_BASE:
      case LEVEL_CLOUD_TOP:
      case LEVEL_ISOTHERM:
      case LEVEL_ADIABAT:
      case LEVEL_MAX_WIND:
      case LEVEL_TROP:
      case LEVEL_MEAN_SEA:
	break;
      case LEVEL_FH:
      case LEVEL_FHG:
	printf("   %24s : %f (M)\n", "Level", mblev(gp->level));
	break;
      case LEVEL_SIGMA:
	printf("   %24s : %f\n", "Level", (256*gp->level[0]+gp->level[1])/10000.0);
	break;
      case LEVEL_DBS:
      case LEVEL_BLS:
      case LEVEL_ISOBARIC:
	printf("   %24s : %f (Pa)\n", "Level", 100.*mblev(gp->level));
	break;
      case LEVEL_LISO:
      case LEVEL_LFHM:
      case LEVEL_LFHG:
      case LEVEL_LS:
      case LEVEL_LBLS:
      case LEVEL_LISH:
      case LEVEL_LSH:
      case LEVEL_LISM:
	printf("   %24s : %f (Pa)\n", "Level 1", gp->level[0]*1000.0);
	printf("   %24s : %f (Pa)\n", "Level 2", gp->level[1]*1000.0);
	break;
    }
    printf("   %24s : %04d/%02d/%02d:%02d:%02d\n","Reference Time", 
	   gp->year+(gp->century - 1)*100,
	   gp->month, gp->day, gp->hour, 
	   gp->minute);
    printf("   %24s : %s\n", "Time Unit", tunitsname(gp->tunit));
    printf("   %24s : %s\n", "Time Range Indicator", triname(gp->tr_flg));
    switch (gp->tr_flg) {
      case TRI_P1:
      case TRI_IAP:
	printf("   %24s : %d\n", "Time 1 (P1)", gp->tr[0]);
	break; 
      case TRI_P12:
      case TRI_Ave:
      case TRI_Acc:
      case TRI_Dif:
      case TRI_LP1:
      case TRI_AvgN:
      case TRI_AvgN1:
      case TRI_AccN1:
      case TRI_AvgN2:
      case TRI_AvgN3:
      case TRI_AccN3:
	printf("   %24s : %d\n", "Time 1 (P1)", gp->tr[0]);
	printf("   %24s : %d\n", "Time 2 (P2)", gp->tr[1]);
	break; 
    }
    printf("   %24s : %d\n", "Decimal Scale Factor", gp->scale10);
    printf("   %24s : %d\n", "Binary Scale Factor", gp->bd->bscale);
    printf("   %24s : %f\n", "Reference Value", gp->bd->ref);
    {				/* all done in float to reproduce FSL value */
	float dscal;
	dscal = gp->scale10;
	dscal = pow(10.0, dscal);
	sprintf(format, "   %%24s : %%.%dg\n", prec);
	printf(format, "Minimum Value", (float)(gp->bd->ref / dscal));
    }
    printf("   %24s : %d\n", "Number of Bits", gp->bd->nbits);
    printf("   %24s : %s\n", "BMS Included", gp->has_bms ? "TRUE" : "FALSE");
    printf("   %24s : %s\n", "GDS Included", gp->has_gds ? "TRUE" : "FALSE");
    printf("   %24s : %s\n", "IsInternationalGrid",
	   is_ixg(gp->grid) ? "TRUE" : "FALSE");
    printf("   %24s : %d\n","GRIB Edition", gp->edition);
    printf("   %24s : %d\n","Parameter Table Ver", gp->param_vers);
    print_gdes(gp->gd);		/* print Grid Description Section */
    if (! headers_only) {
	printf("                 grid values:\n");
        print_floats(gp->data, gp->cols, gp->npts/gp->cols, prec);
    }
}     


static void
print_grib_line(gp)
     struct product_data *gp;
{
    char *lev;			/* abbreviation for level flag */
    
    lev = levelsuffix(gp->level_flg);

    if (lev) {
	printf("%4d%4d%4d%4d ",
	       gp->center, gp->model, gp->grid, gp->param);
    	
	if (lev[0] == '\0')
	    lev = "isob";	/* show isobaric levels explicitly */
	printf("%7s %4d %4d", lev, level1(gp->level_flg, gp->level),
	       level2(gp->level_flg, gp->level));
	
	printf(" %4d%4d%4d %4d%4d%4d %6d %s %4d\n",
	       gp->tr_flg, gp->tr[0], gp->tr[1],
	       gp->bits, gp->has_bms, gp->has_gds, gp->npts, gp->header,
               gp->gd->nv);
    }
}     


/*
 * Parse raw product bytes into product_data structure.  Returns 0 if
 * failed.  User should call free_product_data() on result when done with
 * it.  Also expands quasi-regular grids to full rectangular grids if qmeth
 * is non-null.
 */
static product_data *
grib_decode(prodp, quasp)
     prod *prodp ;		/* input raw product bytes */
     quas *quasp ;		/* if non-null, method used to expand
				   quasi-regular "grids" */
{
    grib1 *gp = new_grib1(prodp); /* overlay raw bits on raw grib1 structure */
    product_data *pdp;

    if (gp == 0)
	return 0;

    pdp = new_product_data(gp);	/* compute cooked product structure, with GDS
				   (manufactured, if necessary) and bytemap */
    free_grib1(gp);

    if (!pdp)
	return 0;

    if (pdp->gd->quasi && quasp) {
	int ret = expand_quasi(quasp, pdp) ; /* Changes *pdp */
	if (!ret) {
	    uerror("%s: can't expand quasi-regular grid", pdp->header);
	    return 0;
	}
    }
    return pdp;
}


static int
do_dump (fp, ep, timeout, prec, quasp)
    FILE *fp;			/* input */
    FILE *ep;			/* if non-null, where to append bad GRIBs */
    int timeout;
    int prec;			/* precision to use for printing floats */
    quas *quasp;		/* if non-null, quasi-reg expansion method */
{
    struct prod the_prod;
    struct product_data *gribp;	/* UPC structure */
    
    num_wmo_messages = 0;
    num_gribs_decoded = 0;

    if (brief_output)
	printf(" cnt mdl grd prm    lvlf lev1 lev2  trf tr0 tr1 bits bms gds   npts header\n");

    while(1) {
	int bytes = get_prod(fp, timeout, &the_prod);
	if (bytes == 0)
	  continue;
	else
	  num_wmo_messages++;
	
	/* decode it into a grib1 product structure */
	gribp = grib_decode(&the_prod, quasp);
	if (gribp == 0) {
	    if (ep) {
		if (fwrite(the_prod.bytes, the_prod.len, 1, ep) == 0) {
		    serror("writing bad GRIB to error file");
		}
	    }
	    continue;
	}
	num_gribs_decoded++;

	/* Dump grib info to stdout */
	if(brief_output)
	    print_grib_line(gribp);
	else
	    print_grib(gribp, prec);
	free_product_data(gribp);
	num_gribs_written++;
    }
}


int
main(ac,av)
     int ac ;
     char *av[] ;
{
    extern int optind;
    char *logfname = "-" ;	/* log file name, default is stdout */
    FILE *fp = stdin;		/* default input */
    FILE *ep = 0;		/* file handle for bad GRIBS output, when
				   -e badfname used */
    int prec = DEFAULT_PRECISION; /* precision to use for printing floats */
    int timeo = DEFAULT_TIMEOUT ; /* timeout */
    quas *quasp = 0;		/* default, don't expand quasi-regular grids */
    /*
     * register exit handler
     */
    if(atexit(cleanup) != 0)
      {
	  serror("atexit") ;
	  exit(1) ;
      }
    
    /*
     * set up signal handlers
     */
    set_sigactions() ;
    
    {
	extern int opterr;
	extern char *optarg;
	int ch;
	int logmask = (LOG_MASK(LOG_ERR) | LOG_MASK(LOG_NOTICE)) ;
	int errflg = 0;
	
	opterr = 1;
	
        headers_only = 0;
        
	while ((ch = getopt(ac, av, "vxhbl:t:e:p:q:g:")) != EOF) {
	    switch (ch) {
	    case 'v':
		logmask |= LOG_MASK(LOG_INFO) ;
		break;
	    case 'x':
		logmask |= LOG_MASK(LOG_DEBUG) ;
		break;
            case 'g': /* gribtonctable */
               fprintf(stderr, "Conversion table in file %s\n",optarg) ;
                user_makeparamtable(optarg);
/*
                user_printparamtable();
*/
                break;
	    case 'l':
		logfname = optarg ;
		break;
	    case 't':
		timeo = atoi(optarg) ;
		if(timeo < 1) {
		    fprintf(stderr, "%s: invalid timeout %s",
			    av[0], optarg) ;
		    errflg++;
		}
		break;
	    case 'h':
		headers_only = 1;
		break;
	    case 'b':
		brief_output = 1;
		headers_only = 1;
		break;
	    case 'e':
		ep = fopen(optarg, "w");
		if(!ep) {
		    serror("can't open %s", optarg);
		    errflg++;
		}
		break;
	    case 'p':
		prec = atoi(optarg) ;
		if(prec < 1 || prec > 99) {
		    fprintf(stderr, "%s: invalid precision %s",
			    av[0], optarg) ;
		    errflg++;
		}
		break;
	    case 'q':
		quasp = qmeth_parse(optarg);
		if(!quasp) {
		    fprintf(stderr,
			    "%s: invalid quasi-regular expansion method %s",
			    av[0], optarg) ;
		    errflg++;
		}
		break;
	    case '?':
		errflg++;
		break;
	    }
	}
	

	if(errflg)
	  usage(av[0]);	
	(void) setulogmask(logmask) ;
    }
        
    /*
     * initialize logger
     */
    (void) openulog(ubasename(av[0]),
		    (LOG_CONS|LOG_PID), LOG_LOCAL0, logfname) ;
    uinfo("Starting Up") ;

    do {
	if (optind < ac) {	/* arguments still left, use as filenames */
	    fp = fopen(av[optind++], "r");
	    if(!fp) {
		serror("can't open %s", av[optind-1]);
		return 1;
	    }
	}
	do_dump(fp, ep, timeo, prec, quasp);
    } while (optind < ac);
    
    return 0;
}

